File: high_level/python/e1_get_available_motions.py
#!/usr/bin/env python3
import grpc
import sys

from proto import grpc_service_pb2
from proto import grpc_service_pb2_grpc


def run_e1(server_address: str = "localhost:50051"):
    """
    Get all available motions from the server. This is useful to understand what motions are available 
    and what parameters they require before executing them.
    
    Args:
        server_address: server address
        fsm_state: optional FSM state filter, empty string returns all motions
    """

    # Establish connection
    channel = grpc.insecure_channel(server_address)
    stub = grpc_service_pb2_grpc.gRPCServiceStub(channel)
    print(f"Connected to server: {server_address}")

    print("Example 1: Get Available Motions")

    # Create request
    request = grpc_service_pb2.GetMotionsRequest()

    response = stub.GetAvailableMotions(request)

    if response.success:
        print(f"Successfully retrieved motion list: {response.message}")
        print(f"Found {len(response.motions)} motions:\n")

        for motion in response.motions:
            motion_id = motion.motion_id
            print(f"  [{motion_id}]")

            # description
            if motion_id in response.descriptions:
                print(f"    Description: {response.descriptions[motion_id]}")

            # some motion parameters
            if motion.parameters:
                print("    Parameters (default values):")
                for param in motion.parameters:
                    print(f"      - {param.key}: ", end="")

                    if param.HasField("float_value"):
                        print(f"{param.float_value} (float)")
                    elif param.HasField("int_value"):
                        print(f"{param.int_value} (int)")
                    elif param.HasField("string_value"):
                        print(f'"{param.string_value}" (string)')
                    elif param.HasField("bool_value"):
                        print(f"{'true' if param.bool_value else 'false'} (bool)")
                    else:
                        print("(not set)")
            print()

    else:
        print(f"\nFailed to retrieve motions: {response.message}")

    channel.close()


if __name__ == "__main__":
    server = sys.argv[1] if len(sys.argv) > 1 else "192.168.5.2:50051"
    run_e1(server)

File: high_level/cpp/e1_get_available_motions.cpp
#include <iostream>
#include <string>
#include <grpcpp/grpcpp.h>
#include "proto/grpc_service.grpc.pb.h"

using grpc_comm::GetMotionsRequest;
using grpc_comm::GetMotionsResponse;
using grpc_comm::gRPCService;

class MotionClient
{
public:
    explicit MotionClient(const std::string& server_address)
        : server_address_(server_address)
    {
        auto channel = grpc::CreateChannel(server_address_, grpc::InsecureChannelCredentials());
        stub_ = gRPCService::NewStub(channel);
    }

    // Query and print all available motions.
    bool Run()
    {
        std::cout << "Connected to server: " << server_address_ << std::endl;
        std::cout << "Example 1: Get Available Motions" << std::endl;

        GetMotionsRequest request;
        GetMotionsResponse response;
        grpc::ClientContext context;

        auto status = stub_->GetAvailableMotions(&context, request, &response);
        if (!status.ok()) {
            std::cout << "RPC failed: " << status.error_message() << std::endl;
            return false;
        }
        if (!response.success()) {
            std::cout << "Failed to retrieve motions: " << response.message() << std::endl;
            return false;
        }

        std::cout << "Successfully retrieved motion list: " << response.message() << std::endl;
        std::cout << "Found " << response.motions_size() << " motions:\n" << std::endl;

        for (const auto& motion : response.motions()) {
            const std::string& motion_id = motion.motion_id();
            std::cout << "  [" << motion_id << "]" << std::endl;

            auto desc_it = response.descriptions().find(motion_id);
            if (desc_it != response.descriptions().end()) {
                std::cout << "    Description: " << desc_it->second << std::endl;
            }

            if (motion.parameters_size() > 0) {
                std::cout << "    Parameters (default values):" << std::endl;
                for (const auto& param : motion.parameters()) {
                    std::cout << "      - " << param.key() << ": ";
                    switch (param.value_case()) {
                        case grpc_comm::Parameter::kFloatValue:
                            std::cout << param.float_value() << " (float)";
                            break;
                        case grpc_comm::Parameter::kIntValue:
                            std::cout << param.int_value() << " (int)";
                            break;
                        case grpc_comm::Parameter::kStringValue:
                            std::cout << '"' << param.string_value() << '"' << " (string)";
                            break;
                        case grpc_comm::Parameter::kBoolValue:
                            std::cout << (param.bool_value() ? "true" : "false") << " (bool)";
                            break;
                        default:
                            std::cout << "(not set)";
                            break;
                    }
                    std::cout << std::endl;
                }
            }
            std::cout << std::endl;
        }
        return true;
    }

private:
    std::unique_ptr<gRPCService::Stub> stub_;
    std::string server_address_;
};

int main(int argc, char** argv)
{
    const std::string server_address = (argc > 1) ? argv[1] : "192.168.5.2:50051";
    MotionClient client(server_address);
    return client.Run() ? 0 : 1;
}
